main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver2.scala

```diff
@@ -0,0 +1,580 @@
+/*
+ * Copyright 2024 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.typer
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.Ast.{Instance, TraitContext}
+import ca.uwaterloo.flix.language.ast.shared.Scope
+import ca.uwaterloo.flix.language.ast.{Ast, Kind, RigidityEnv, SourceLocation, Symbol, Type}
+import ca.uwaterloo.flix.language.phase.typer.TypeReduction2.reduce
+import ca.uwaterloo.flix.language.phase.unification.*
+import ca.uwaterloo.flix.util.collection.{ListMap, MapOps}
+import ca.uwaterloo.flix.util.{InternalCompilerException, Result}
+
+/**
+  * The constraint solver reduces a collection of constraints by iteratively applying reduction rules.
+  * The result of constraint solving is a substitution and a list of constraints that could not be resolved.
+  */
+object ConstraintSolver2 {
+
+  sealed trait TypeConstraint
```

Comment:
What do you say to moving TypeConstraint, Progress, Soup, SubstitutionTree into their own files?

Reply:
Yes

Reply:
Except Soup, which is linked to what goes on here.